stdlib/src/builtin/simd.mojo

```diff
@@ -1939,6 +1939,96 @@ struct SIMD[type: DType, size: Int](
+
+        It's done using the specified mask (permutation). The mask
+        values must be within `len(self)`. If that's not the case,
+        the behavior is undefined.
+
+        The mask is not known at compile time, unlike the `shuffle` method.
+
+        Note that currently, this function is fast only if the following
+        conditions are met:
+        1) The SIMD vector `self` is of type uint8 and size 16
+        2) The CPU supports SSE4 or NEON
+
+        If that's not the case, the function will fallback on a slower path,
+        which is an unrolled for loop.
+
+        The pseudocode of this function is:
+        ```mojo
+        var result = SIMD[Self.type, mask_size]()
+        for i in range(0, mask_size):
+            result[i] = self[int(mask[i])]
+        ```
+
+        Parameters:
+            mask_size: The size of the mask.
+
+        Args:
+            mask: The mask to use. Contains the indices to use to shuffle.
+
+        Returns:
+            A new vector with the same length as the mask where the value at
+            position `i` is equal to `self[mask[i]]`.
+        """
+
+        @parameter
+        if (
+            (sys.has_sse4() or sys.has_neon())
+            and Self.type == DType.uint8
+            and Self.size == 16
+        ):
+            # The instruction works with mask size of 16
+            alias target_mask_size = 16
+
+            # We know that simd sizes are powers of two, so we can use recursivity
+            # to iterate on the method until we reach the target size.
+            @parameter
+            if mask_size < target_mask_size:
+                # Make a bigger mask and retry
+                var new_mask = mask.join(SIMD[DType.uint8, mask_size]())
+                return self.dynamic_shuffle(new_mask).slice[mask_size]()
+            elif mask_size == target_mask_size:
+                # The compiler isn't smart enough yet. It complains about parameter mismatch
+                # because it cannot narrow them. Let's help it a bit and hope it
+                # understands this is a no-op. So far there wasn't any performance
+                # drop likely indicating that the compiler is doing the right thing.
+                # Everything commented with no-op should not generate any instruction.
+                var indices_copy = mask.slice[16, offset=0]()  # no-op
+                var self_copy = self.slice[16, offset=0]().cast[
+                    DType.uint8
+                ]()  # no-op
+                var result = _pshuf_or_tbl1(self_copy, indices_copy)
+                return result.slice[mask_size]().cast[Self.type]()  # no-op
+            elif mask_size > target_mask_size:
+                # We split it in two and call dynamic_shuffle twice.
+                var first_half_of_mask = mask.slice[mask_size // 2, offset=0]()
+                var second_half_of_mask = mask.slice[
+                    mask_size // 2, offset = mask_size // 2
+                ]()
```

Comment:
Since this involves a tuple, it may impact codegen. Since this is for a performance sensitive work (utf8 validatation), I'd prefer that we change it after we have some solid automatic benchmarking infrastructure in place. While working on utf-8 validation, I noticed that the most insignificant changes could impact codegen so I'm a bit worried about modifying things.

Reply:
Fair, plus we could always land and iterate.